contents/scala/Scala.md

```diff
@@ -0,0 +1,241 @@
+
+Following are some examples to illustrate the benefit of Scala:
+
+## Operator Overloading
+All operators in Scala such as +, -, *, /. are actually methods.
+
+Normally, we write addition in following way:
+
+`val a = 1 + 2`
+
+However, in Scala, it will be interpreted as following:
+
+`val a = 1.+(2)`
+
+Knowing that, we are able to overload operators just like method overloading. It allows us to redefine the behavior of
+arithmetic operators and give them meaning for our own custom classes.
+
+Following is an elegant *ComplexInt* class which uses **operator overloading**.
+
+```
+case class ComplexInt(real: Int, im: Int) {
+  def + (other: ComplexInt) = ComplexInt(real + other.real, im + other.im)
+  def * (other: ComplexInt) = ComplexInt(
+    real * other.real - im * other.im,
+    real * other.im + im * other.real
+  )
+  def unary_- = ComplexInt(-real, -im)
+  def - (other: ComplexInt) = this + -other
+
+  override def toString = real + " + " + im + "i"
+}
+```
+
+We are able to use the class like below:
+
+```
+val a = ComplexInt(1, 1)
+val b = ComplexInt(1, -1)
+val c = a * b
+val d = -c + a
+```
+
+In other languages like Java, the similar operations of the same class may need to performed like following:
+
+```
+ComplexInt a = new ComplexInt(1, 1);
+ComplexInt b = new ComplexInt(1, -1);
+ComplexInt c = a.times(b);
+ComplexInt d = (c.negative()).add(a);
+```
+
+
+## Mixed-In Trait
+In Scala, class abstractions are extended by subclassing and by a flexible mixin-based composition mechanism to 
+avoid the problems of multiple inheritance. 
+
+Following is the piece of code to show the uniqueness of OOP in Scala:
+```
+abstract class ParentTrait {
+  // abstract
+  def print()
+}
+
+class Sample extends ParentTrait {
+  override def print() {
+    println("print in Sample")
+  }
+}
+
+trait TraitA extends ParentTrait {
+  abstract override def print() {
+    println("print in TraitA")
+    super.print()
+  }
+}
+
+trait TraitB extends ParentTrait {
+  abstract override def print() {
+    println("print in TraitB")
+    super.print()
+  }
+}
+
+trait TraitC extends ParentTrait {
+  abstract override def print() {
+    println("print in TraitC")
+    super.print()
+  }
+}
+
+object Sample {
+  def main(args: Array[String]): Unit = {
+    val example = new Sample with TraitA with TraitB with TraitC
+   example.print()
+  }
+}
+```
+The following is the execution result of above code:
+```
+print in TraitC
+print in TraitB
+print in TraitA
+print in Sample
+```
+The explanation is that the call to `print()` firstly executes the code in `TraitC`, which is the last trait mixed in.
+Then through the `super()` calls, it threads back through the other mixed-in traits. And eventually to the code in 
+`Sample`. Even though none of the traits inherited from one another, it works properly. 
+
+This is similar to the decorator pattern but is more *concise* and less *error-prone*. In other languages, a similar 
```

Comment:
Similar here. The disadvantage part can be move at the beginning, and provide some short but clear example of disadvantage. If readers know what's the problem, they may understand the features in Scala better.